File: leetcode/0662-maximum-width-of-binary-tree/solution.py
from typing import Optional
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        max_width = 0
        queue = deque([(root, 0)])

        while queue:
            level_size = len(queue)
            _, first_index = queue[0]
            last_index = first_index

            for _ in range(level_size):
                node, index = queue.popleft()
                index -= first_index  # re-base to avoid overflow
                last_index = index

                if node.left:
                    queue.append((node.left, 2 * index))
                if node.right:
                    queue.append((node.right, 2 * index + 1))

            max_width = max(max_width, last_index + 1)

        return max_width

File: leetcode/0662-maximum-width-of-binary-tree/Solution.java
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private static class Item {
        TreeNode node;
        long index;

        Item(TreeNode node, long index) {
            this.node = node;
            this.index = index;
        }
    }

    public int widthOfBinaryTree(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int maxWidth = 0;
        Deque<Item> queue = new ArrayDeque<>();
        queue.add(new Item(root, 0));

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            long firstIndex = queue.peek().index;
            long lastIndex = firstIndex;

            for (int i = 0; i < levelSize; i++) {
                Item item = queue.poll();
                long index = item.index - firstIndex;
                lastIndex = index;

                if (item.node.left != null) {
                    queue.add(new Item(item.node.left, 2 * index));
                }
                if (item.node.right != null) {
                    queue.add(new Item(item.node.right, 2 * index + 1));
                }
            }

            maxWidth = (int) Math.max(maxWidth, lastIndex + 1);
        }

        return maxWidth;
    }
}
